### PR TITLE
Allow use of custom delimiter for paths in module generator

### PR DESCRIPTION
**Pull request by Micket**
This is the straight forward option that we definitely should expose.
It would allow any easyblock to use this at least, though in practice, we should also enable some way for easyconfigs to chose this when specifying modextrapaths.

I left that out for this initial PR, since we need to introduce some new syntax. Maybe good to mark that for 5.0.x.

Various ideas have been thrown out:

```python
modextrapaths = {
    'PATH': 'some_path',    # Normal : delimited paths continue to be either a string
    'PATH': ['some_path', 'other_path'],    # or a list of strings
    'FOO': ('foo', ';'), # Using a tuple indicates second argument is delimiter ; in this case
    'FOO': (['foo', 'bar'], ';'), # same with multiple paths, the string is replaced by a list.
    'BAR': {'paths': 'foo', 'delimiter': ';'}, # more explicit option
    'BAR': {'paths': ['foo', 'bar'], 'delimiter': ';'}, # and with multiple strings
}
```

That's about all the options I can think of. I would be against doing

```python
modextrapaths = {
    'FOO': [('foo', ';'), ('bar', ';')], # not good
}
```

since it could allow for different separators to be used for the same variable, which doesn't make any sense.

Needs tests.

### CODE REVIEW

**Comment by jfgrimm** [comment]
personally, I'd be in favour of the more explicit:
```
    'BAR': {'paths': 'foo', 'delim': ';'}, # more explicit option
    'BAR': {'paths': ['foo', 'bar'], 'delim': ';'}, # and with multiple strings
```

**Comment by jfgrimm** [comment]
```
FAIL: test_make_module_step (test.framework.easyblock.EasyBlockTest)
Test the make_module_step
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/tmp/runner/1f60854bac9a4cfba15aca0717bfcfe964c33e1f/lib/python3.6/site-packages/test/framework/easyblock.py", line 1346, in test_make_module_step
    self.assertTrue(regex.search(txt), "Pattern %s found in %s" % (regex.pattern, txt))
AssertionError: None is not true : Pattern ^prepend-path\s+(-d ".")?TCLLIBPATH\s+\$root/paths$ found in #%Module
proc ModulesHelp { } {
    puts stderr {

Description
===========
This \{is a\}\} \[fancy\]\] \[\[description\]\]. \{\{\[\[TEST\}\]


More information
================
 - Homepage: http://example.com/
    }
}

module-whatis {Description: This \{is a\}\} \[fancy\]\] \[\[description\]\]. \{\{\[\[TEST\}\]}
module-whatis {Homepage: http://example.com}/
module-whatis {URL: http://example.com}/

set root /tmp/eb-2q3i9mdp/eb-evnuh1i2/eb-epx33r9k/tmpu4hfzyac/software/pi/3.14

conflict pi

module load GCC/6.4.0-2.28

module load test/.1.2.3

prepend-path	CMAKE_PREFIX_PATH		$root
prepend-path	PATH		$root/bin
setenv	EBROOTPI		"$root"
setenv	EBVERSIONPI		"3.14"
setenv	EBDEVELPI		"$root/easybuild/pi-3.14-easybuild-devel"

setenv	PI		"3.1415"
setenv	FOO		"bar"
pushenv	TEST_PUSHENV		"123"
prepend-path	PATH		$root/xbin
prepend-path	PATH		$root/pibin
prepend-path	CPATH		$root/pi/include
prepend-path -d " "	TCLLIBPATH		$root/pi 
append-path	APPEND_PATH		$root/append_path
# Built with EasyBuild version 5.0.0.dev0
```

**Comment by Micket** [comment]
had a 2 day meeting that got in the way of finishing this. I hate regexes. The generated file should be OK i think. Lets see if i got it right now

File: easybuild/framework/easyblock.py [hunk 1]
```diff
@@ -1469,21 +1469,29 @@ def make_module_extra(self, altroot=None, altversion=None):
...
+                    if not isinstance(paths, (list, str)):
+                        raise EasyBuildError('modextrapaths dict value "{value}" paths must be list or str')
+                    if not isinstance(delim, str):
+                        raise EasyBuildError('modextrapaths dict value "{value}" delimiter must be a str')
+                    lines.append(self.module_generator.update_paths(key, paths, prepend=prepend, delim=delim,
+                                                                    allow_abs=allow_abs))
+                else:
+                    if isinstance(value, str):
+                        value = [value]
+                    lines.append(self.module_generator.update_paths(key, value, prepend=prepend, allow_abs=allow_abs))
```
Comment (lexming) [inline easybuild/framework/easyblock.py:1495]: This seems overly complex with too many type checks. Many of these checks are already done by `update_paths`, so I think we can simplify to:

```suggestion
        for extrapaths_type, prepend in [('modextrapaths', True), ('modextrapaths_append', False)]:
            allow_abs = self.cfg['allow_prepend_abs_path'] if prepend else self.cfg['allow_append_abs_path']

            for (key, value) in self.cfg[extrapaths_type].items():
                if not isinstance(value, (tuple, list, dict, str)):
                    raise EasyBuildError(
                        f"{extrapaths_type} dict value '{value}' (type {type(value)}) is not a 'list, dict or str'"
                    )

                try:
                    paths = value['paths']
                    delim = value['delimiter']
                except (KeyError, TypeError):
                    paths = value
                    delim = ':'

                lines.append(
                    self.module_generator.update_paths(key, paths, prepend=prepend, delim=delim, allow_abs=allow_abs)
                )

```

These code passes unit tests.

**Review by lexming — CHANGES_REQUESTED** [review]
Looks good, I only have some comments to simplify the code.

File: easybuild/tools/module_generator.py [hunk 2]
```diff
@@ -1010,7 +1016,10 @@ def update_paths(self, key, paths, prepend=True, allow_abs=False, expand_relpath
             else:
                 abspaths.append(path)
 
-        statements = ['%s-path\t%s\t\t%s\n' % (update_type, key, p) for p in abspaths]
+        if delim != ':':
+            statements = ['%s-path -d "%s"\t%s\t\t%s\n' % (update_type, delim, key, p) for p in abspaths]
+        else:
+            statements = ['%s-path\t%s\t\t%s\n' % (update_type, key, p) for p in abspaths]
```
[the remark this answers is not in the document]
Comment (boegel) [reply easybuild/tools/module_generator.py:1022]: suggestion:

```python
delim_opt = '' if delim == ':' else f'-d "{delim}"'
```

and then use `delim_opt` in `statements`

File: easybuild/framework/easyblock.py (hunk 1 above)
Comment (Micket) [reply easybuild/framework/easyblock.py:1495]: well ok, though I must admit I don't love the reliance on exception types to implicitly determine the type.
The update_paths part seems unclear what is supposed to take (I would actually prefer if that code instead just took a list like the API docs says, and deal with sanitizing user inputs as early as possible)

**Review by lexming — APPROVED** [review]
LGTM

File: easybuild/tools/module_generator.py [hunk 3]
```diff
@@ -1010,7 +1016,8 @@ def update_paths(self, key, paths, prepend=True, allow_abs=False, expand_relpath
             else:
                 abspaths.append(path)
 
-        statements = ['%s-path\t%s\t\t%s\n' % (update_type, key, p) for p in abspaths]
+        delim_opt = '' if delim != ':' else f' -d "{delim}"'
```
Comment (lexming) [inline easybuild/tools/module_generator.py:1019]:
```suggestion
        delim_opt = '' if delim == ':' else f' -d "{delim}"'
```

**Review by lexming — CHANGES_REQUESTED** [review]
LGTM but needs a minor fix

File: easybuild/tools/module_generator.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (Micket) [reply easybuild/tools/module_generator.py]: I didn't want to add that extra noise into all the module files for something that will be extremely rarely used (something like 3 softwares)

[the remark this answers is not in the document]
Comment (lexming) [reply easybuild/tools/module_generator.py]: ok, fair enough :slightly_smiling_face:

**Review by lexming — APPROVED** [review]
LGTM

**Pull request merged**